internal/app/app.go

```
@@ -0,0 +1,153 @@
+	}
+
+	err = p.Apply()
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	return nil
+}
+
+// New creates a new App object.
+func New(opts ...option) (a *App, err error) {
+	defer decorate.OnError(&err, "cannot initialize application")
+
+	// defaults
+	args := options{
+		timeout: 1 * time.Second,
+		initCtx: context.Background(),
+	}
+	// applied options
+	for _, o := range opts {
+		o(&args)
+	}
+
+	// Don’t call dbus.SystemBus which caches globally system dbus (issues in tests)
+	conn, err := dbus.ConnectSystemBus()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancelCause(args.initCtx)
+	obj := busObject{
+		ctx:    ctx,
+		cancel: cancel,
+	}
+
+	if err = conn.Export(&obj, dbusObjectPath, dbusInterface); err != nil {
+		_ = conn.Close()
+		return nil, err
+	}
+	if err = conn.Export(introspect.NewIntrospectable(&introspect.Node{
+		Name: dbusObjectPath,
+		Interfaces: []introspect.Interface{
+			introspect.IntrospectData,
+			{
+				Name:    dbusInterface,
+				Methods: introspect.Methods(&obj),
+			},
+		},
+	}), dbusObjectPath, introspect.IntrospectData.Name); err != nil {
+		_ = conn.Close()
+		return nil, err
+	}
+
+	reply, err := conn.RequestName(dbusInterface, dbus.NameFlagDoNotQueue)
+	if err != nil {
+		_ = conn.Close()
+		return nil, err
+	}
+	if reply != dbus.RequestNameReplyPrimaryOwner {
+		_ = conn.Close()
+		return nil, fmt.Errorf("D-Bus name already taken")
+	}
+
+	return &App{
+		conn:    conn,
+		obj:     &obj,
+		timeout: args.timeout,
+	}, nil
+}
+
+// Wait blocks until the context is cancelled or the timeout is reached,
+// returning an appropriate exit code and closing the D-Bus connection before
+// exiting.
+func (a *App) Wait() int {
+	defer a.conn.Close()
+
+	for {
+		select {
+		case <-time.After(a.timeout):
+			log.Infof("No signal received, exiting")
+			return 0
+		case <-a.obj.ctx.Done():
+			if err := context.Cause(a.obj.ctx); err != nil && !errors.Is(err, context.Canceled) {
+				log.Error(err)
```

I would rather return an error here, and log in the caller.

---

it’s Friday, I can take it :p